Feature/pathfinding

Banananaman91 opened the pull request:
Begun creating pathfinding algorithm

lizziebriggs committed:
Started player movement

Created a player movement script

Banananaman91 committed:
Sorted .gitignore

I made a boo boo and fixed the .gitignore

lizziebriggs committed:
Completed player controls

Player now only moves along adjacent tiles which the user clicks on to move the player to that specific tile

Banananaman91 committed:
Merge remote-tracking branch 'origin/feature/Player-Movement' into feature/Player-Movement

Banananaman91 committed:
Changes to player controller

Change the way the player controller functions. Currently not fully working, needs more work.

Banananaman91 committed:
Begun PathFinding algorithm

Begun creating the PathFinding algorithm. Reorganised files and folders.

Banananaman91 committed:
Removed unused game objects from scene

Banananaman91 committed:
Merging development into player movement

Banananaman91 committed:
New MouseSelection type from development branch

Banananaman91 committed:
Merged player-movement to pathfinding

Banananaman91 committed:
Removed duped types from merging

Banananaman91 committed:
Fixed errors from merging

Banananaman91 committed:
Bugfix endless loop

fixed bug in PathFinder that generated an endless loop.

lizziebriggs committed:
Minor tweak to path finding

Banananaman91 committed:
Fixes to pathfinder

fixed unity crash for pathfinder, turned it into a coroutine, and altered some syntax. Currently still does not move towards target. F values not working properly.

lizziebriggs committed:
Neatened code

Minor tweaks to the code to neaten it up and moved the calculation for the rawGridPoint to a seperate method. No new problems created.

lizziebriggs committed:
Fixed pathfinding algorithm

Fixed the pathfinding algorithm and began creating a method to create a path the player follows.

lizziebriggs committed:
Player moving cont.

Player can move but unity crashes when you try to move more than once

Banananaman91 committed:
Changes to pathFinder return value

Path Finder now returns list value for use, debug shows it working but player still isn't moving

lizziebriggs committed:
player movement cont.

Banananaman91 committed:
Cleaned up code

Cleaned up code, removed commented code. PathFinder appears to mostly generate paths to the right of the player (in scene)

Banananaman91 committed:
Added .gitkeep for empty folders

Banananaman91 committed:
Completed Path Finder

path finder finished, currently returns to player for testing and shows correct path drawn out with movement. Ready to be used for proper player movement and AI movement

Banananaman91 committed:
Final adjustments to Path Finder

Path finder has been adjusted to work for 2D and 3D path finding, to be used across turn based AI and free roaming AI

Banananaman91 committed:
Merge pull request #41 from Banananaman91/feature/pathfinding

Feature/pathfinding

DatPinkGuy committed:
Visible character movement & Precise waypoint selection

Added visible character movement to the waypoint, tile is created now in the position of mouse click, second click on tile commences the movement

Banananaman91 committed:
Merge pull request #43 from Banananaman91/feature/pathfinding

Visible character movement & Precise waypoint selection

Banananaman91 committed:
Put pathfinder behind interface

Created interface for pathfinding, created new messenger system for handling initialisation of pathfinder

Banananaman91 committed:
Merge pull request #44 from Banananaman91/feature/pathfinding

Put pathfinder behind interface

DatPinkGuy committed:
Player Model Placeholder, Turning animation

Added placeholder for player model, creating turning animation during movement to different positions

Banananaman91 committed:
Object avoidance for path finder

Path finding now takes in list of positions to be avoided based on objects in a radius. This also ignores the tile generated for the target position

DatPinkGuy committed:
Added visualized path to Target Position

Visualized path to target position which appears on mouse click and disappears upon reaching target position. Added material to add transparency to the path created.

Banananaman91 committed:
Merge branch 'feature/pathfinding' into feature/Player-Movement

Banananaman91 committed:
Merge pull request #45 from Banananaman91/feature/Player-Movement

Feature/player movement

Banananaman91 committed:
Created controller listener

Banananaman91 committed:
Fixed controller using pathfinder

Banananaman91 committed:
Created object positioner type

This type repositions standard objects to whole values

Banananaman91 committed:
Object avoidance working

Object avoiding fully works, path no longer draws through larger objects. Currently lacks efficiency, performance takes a big hit.

Banananaman91 committed:
Moved object avoidance from player to pathfinder

Pathfinder now contains object avoidance method, usable by all that use pathfinder

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace TurnBasedAssets.Scripts.Controllers
+{
+    public class Controller : MonoBehaviour
```
Empty type? Any reason for this being here?

Banananaman91 replied:
Currently any controllers inherit from this type, including an object controller. The function is to have them all handle re positioning themselves automatically to the correct axis, which will also extend to all enemy AI. Having them inherit from this type allowed me to have them all send themselves as the argument and it fulfilling a single argument parameter.

Banananaman91 replied:
Made this class no longer empty

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/ObjectPositioner.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    public class ObjectPositioner : Controller
+    {
+        private IPosition _iPosition;
```
id rename this to `_position`.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/PositionController.cs:
```diff
@@ -0,0 +1,14 @@
+    public class PositionController : IPosition
+    {
+        private Vector3 _newPosition;
+        public Vector3 RePosition(Vector3 position, Vector3 planePosition)
```
`Reposition` is one word - I'd advise applying the naming accordingly.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Interface/IPathFinder.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace TurnBasedAssets.Scripts.Interface
+{
+    public interface IPathFinder
```
`IPathfinder` would be a more appropriate name as pathfinder is a single word.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Interface/IPathFinder.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public interface IPathFinder
+    {
+        IEnumerator FindPath(Vector3 startPosition, Vector3 targetPosition, bool is3D, float movementRadius, Action<IEnumerable<Vector3>> onCompletion);
```
`is3d` would be more appropriate than `is3D` as per your naming conventions.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MessageBroker/MessageBroker.cs:
```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Collections.Generic;
+using TurnBasedAssets.Scripts.PathFinding;
```
should fix the namespace to be Pathfinding rather than PathFinding.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MessageBroker/PathFinderRequestMessage.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        public PlayerController RequestingComponent { get; }
+
+        public PathFinderRequestMessage(PlayerController requestingComponent)
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MessageBroker/PositionControllerRequestMessage.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public Controller RequestingComponent { get; }
+
+        public PositionControllerRequestMessage(Controller requestingComponent)
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+{
+    public class MouseSelection : MonoBehaviour
+    {
+        private ISelection _selection;
```
Consider reshuffling these fields so public, private and serialized are all grouped and organised. Would make it easier on the eyes.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+        [SerializeField] private Camera _mainCamera;
+        private Vector3 _rawGridPoint;
+        private Vector3 _previousGridPoint;
+        [SerializeField] private float movableRadius;
```
This field is breaking naming conventions.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+        private Vector3 _rawGridPoint;
+        private Vector3 _previousGridPoint;
+        [SerializeField] private float movableRadius;
+        [SerializeField] private float distanceY;
```
This field is breaking naming conventions.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+        private Vector3 _previousGridPoint;
+        [SerializeField] private float movableRadius;
+        [SerializeField] private float distanceY;
+        [SerializeField] private float offset;
```
This field is breaking naming conventions.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+        [SerializeField] private float movableRadius;
+        [SerializeField] private float distanceY;
+        [SerializeField] private float offset;
+        [SerializeField] private float gridSize;
```
This field is breaking naming conventions.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+        [SerializeField] private float distanceY;
+        [SerializeField] private float offset;
+        [SerializeField] private float gridSize;
+        [SerializeField] private GameObject selectedTile;
```
This field is breaking naming conventions.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+        private void Update()
+        {
+            Ray ray = Camera.main.ScreenPointToRay(Input.mousePosition);
+            ;
```
someone let the cat walk on the keyboard? :D

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+        {
+            Ray ray = Camera.main.ScreenPointToRay(Input.mousePosition);
+            ;
+            RaycastHit hit;
```
Aren't you in C# 7.3 now? No reason to use out the C# 4 way here I don't think?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PathFinding/PathFinder.cs:
```diff
@@ -0,0 +1,170 @@
+            }
+        }
+
+        public IEnumerator FindPath(Vector3 startPosition, Vector3 targetPosition, bool is3D, float movementRadius, Action<IEnumerable<Vector3>> onCompletion)
```
Consider refactoring this into multiple methods. While the code fits conventions its still a little hard to read with all the functionality this one method has.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PathFinding/PathFinder.cs:
```diff
@@ -0,0 +1,170 @@
+
+                foreach (var adjacentSquare in adjacentSquares)
+                {
+                    if (closedList.Any(x => x.PositionInWorld == adjacentSquare.PositionInWorld)) // If the adjacentSqaure is in the closedList
```
Couldn't this be a braceless one line if?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PathFinding/PathFinder.cs:
```diff
@@ -0,0 +1,170 @@
+                        continue;
+                    }
+
+                    if (openList.All(x => x.PositionInWorld != adjacentSquare.PositionInWorld)) // If the adjacentSquare is not in the openList
```
Couldn't this be a braceless one line if?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PathFinding/PathFinder.cs:
```diff
@@ -0,0 +1,170 @@
+                        openList.Add(adjacentSquare);
+                    }
+
+                    else if (adjacentSquare.F < openList.First(x => x.PositionInWorld == adjacentSquare.PositionInWorld).F)
```
Couldn't this be a braceless one line if?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PlayerControls/PlayerController.cs:
```diff
@@ -0,0 +1,120 @@
+        public IEnumerator StartPlayerMovement()
+        {
+            mouseSelectionScript.enabled = false;
+            foreach (var LOCATION in path)
```
SCREAMING_CASE variable name is a big no no. `location` would be a better name.

RubyNova reviewed (changes requested):
Overall its mostly just typos I want changed and a bit of refactoring to be done. Other than that nothing really major to be honest.

Banananaman91 committed:
Code review changes

Made appropriate changes based on code review

Banananaman91 committed:
Scene changes

Saved changes within unity scene

Banananaman91 committed:
Merge pull request #46 from Banananaman91/feature/pathfinding

Feature/pathfinding

Banananaman91 committed:
Minor change to using namespace

Banananaman91 committed:
Minor changes, added terrain to scene

Mouse Selection needs redesign to detect terrain that blocks plane, otherwise still being clickable if below

Banananaman91 committed:
Cleaned up code

Banananaman91 committed:
Revamped object avoidance

Changed how object avoidance worked to be more user friendly, objects now handle themselves against pathfinder. Users still need to add method calls at start. All objects using this must inherit from Controller type. Currently still doesn't work with terrain

Banananaman91 committed:
Cleaned up for code review and merging

Cleaned up code for review and cleared scene of unnecessary objects ready for merging

lizziebriggs committed:
Started NPC movement

Just added in a model for an NPC

lizziebriggs committed:
Delete NPCControls.meta

Banananaman91 committed:
Fixed conflicts

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+        internal IPathfinder Pathfinder;
+        private IPosition _position;
+        private ObjectAvoidance _avoidance;
+        [SerializeField] internal MouseSelection mouseSelectionScript;
```
Shouldn't this be `protected` and not `internal`?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+        private IPosition _position;
+        private ObjectAvoidance _avoidance;
+        [SerializeField] internal MouseSelection mouseSelectionScript;
+        [SerializeField] internal GameObject pathFinderTiles;
```
Shouldn't this be `protected` and not `internal`?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+        private ObjectAvoidance _avoidance;
+        [SerializeField] internal MouseSelection mouseSelectionScript;
+        [SerializeField] internal GameObject pathFinderTiles;
+        [SerializeField] internal float movementSpeed;
```
Shouldn't this be `protected` and not `internal`?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+        [SerializeField] internal MouseSelection mouseSelectionScript;
+        [SerializeField] internal GameObject pathFinderTiles;
+        [SerializeField] internal float movementSpeed;
+        [SerializeField] internal float rotationSpeed;
```
Shouldn't this be `protected` and not `internal`?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+        [SerializeField] internal GameObject pathFinderTiles;
+        [SerializeField] internal float movementSpeed;
+        [SerializeField] internal float rotationSpeed;
+        public Renderer RenderBounds => GetComponent<Renderer>();
```
I recommend caching this if you rely on this property frequently - its a minor optimisation but `GetComponent<T>` is more expensive than returning a backing field, naturally.

Banananaman91 replied:
How would you cache this? This is used primarily by pathfinder, checking if a space is within the bounds

RubyNova replied:
I'd just make a private field and assign it, something like...
```
public Renderer RenderBounds => _rendererBounds ?? _rendererBounds = GetComponent<Renderer>();

Banananaman91 replied:
If it was done this way, this private field would never be used. Unless you mean to be Serialized?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+        private void GetPathfinder()
+        {
+            //uses message broker to return pathfinder delegate
+            MessageBroker.MessageBroker.Instance.SendMessageOfType(new PathFinderRequestMessage(this));
```
Why is this `MessageBroker.MessageBroker.Instance` instead of `MessageBroker.Instance`?

Banananaman91 replied:
I couldn't find the reason why, but it will only work this way

RubyNova replied:
I advise investigating this then as it's pretty weird.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+            AvoidMe();
+        }
+
+        private void GetPathfinder()
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+            MessageBroker.MessageBroker.Instance.SendMessageOfType(new PathFinderRequestMessage(this));
+        }
+
+        private void GetObjectReposition()
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+            MessageBroker.MessageBroker.Instance.SendMessageOfType(new PositionControllerRequestMessage(this));
+        }
+
+        private void AvoidMe()
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+            _avoidance.Objects.Add(this);
+        }
+
+        private void AddToObjectAvoidance()
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+            MessageBroker.MessageBroker.Instance.SendMessageOfType(new ObjectRequestMessage(this));
+        }
+
+        public void ObjectInitialise(ObjectAvoidance objectAvoidance)
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+            _avoidance = objectAvoidance;
+        }
+        
+        public void PositionInitialise(IPosition position)
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+            _position = position;
+        }
+        
+        public void PathInitialise(IPathfinder pathfinder)
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/Controller.cs:
```diff
@@ -0,0 +1,72 @@
+            Pathfinder = pathfinder;
+        }
+        
+        private void SetPosition()
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/ObjectPositioner.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace TurnBasedAssets.Scripts.Controllers
+{
+    public class ObjectPositioner : Controller
```
Empty type is empty. Maybe you should remove this?

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/Controllers/PositionController.cs:
```diff
@@ -0,0 +1,14 @@
+    public class PositionController : IPosition
+    {
+        private Vector3 _newPosition;
+        public Vector3 Reposition(Vector3 position, Vector3 planePosition)
```
You can expression body this.

Banananaman91 replied:
Your recommendation creates errors for the return value

RubyNova replied:
Then you're inputting it incorrectly I guess? `=> _newPosition = new Vector3(Mathf.Round(position.x), planePosition.y, Mathf.Round(position.z));` is legal code.

Banananaman91 replied:
It was inputted incorrectly, thank you

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MessageBroker/MessageBroker.cs:
```diff
@@ -0,0 +1,38 @@
+    {
+        private Dictionary<Type, List<Delegate>> _messageMappings;
+        
+        public MessageBroker()
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MessageBroker/MessageBroker.cs:
```diff
@@ -0,0 +1,38 @@
+        public void RegisterMessageOfType<T>(Action<T> messageHandler)
+        {
+            var messageType = typeof(T);
+            if(!_messageMappings.ContainsKey(messageType))
```
You could one-line this if statement if you like.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+                            _selection = hit.collider.GetComponent<ISelection>();
+                            _selection.Select();
+                        }
+
```
Random gap here. Consider removing.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/MouseController/MouseSelection.cs:
```diff
@@ -0,0 +1,95 @@
+                            _selection = newTile.GetComponent<ISelection>();
+                        }
+                    }
+
```
Random gap here. Consider removing.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PathFinding/ObjectAvoidance.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        public List<Controller> Objects = new List<Controller>();
+
+        public void UpdateSpaces(Controller controller)
```
You can expression body this.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PathFinding/ObjectAvoidance.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class ObjectAvoidance : IAvoider
+    {
+        public List<Controller> Objects = new List<Controller>();
```
Public fields contradict your codebase's conventions. Please revise.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PathFinding/Pathfinder.cs:
```diff
@@ -0,0 +1,146 @@
+        private List<Vector3> _pathToFollow = new List<Vector3>();
+        private float _2dMaxDistance = 1;
+        private ObjectAvoidance _avoidance;
+        public class Location
```
This doesn't need to be nested. Please refactor this away from your Pathfinder type.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PathFinding/Pathfinder.cs:
```diff
@@ -0,0 +1,146 @@
+        {
+            List<Location> returnList = new List<Location>();
+
+            for (float xIndex = point.PositionInWorld.x - 1; xIndex <= point.PositionInWorld.x + 1; xIndex++)
```
While I understand why you've used a float as an iterator in this context it does read a little strange. For future note I advise using `int`. There's no point refactoring this now however as I can see it'd cause too many changes.

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/PlayerControls/PlayerController.cs:
```diff
@@ -0,0 +1,83 @@
+        private List<GameObject> _pathVisualized = new List<GameObject>();
+        private Vector3 _previousLocation;
+        private Vector3 _previousDistance;
+        private float SphereRadius => mouseSelectionScript.MovableRadius;
```
That's a clever use of a property to essentially alias that value though private properties are not the norm. No need to refactor this, just leaving a remark.

Banananaman91 replied:
Thank you

RubyNova commented on GameProject/Assets/TurnBasedAssets/Scripts/SelectedTile/SelectedTile.cs:
```diff
@@ -0,0 +1,35 @@
+            StartCoroutine(_playerController.FindPossibleMovePositions(_mouseSelection.RawGridPoint));
+        }
+
+        public void Select()
```
You can expression body this.

RubyNova reviewed (changes requested):
Just mainly smaller changes and remarks on the codebase generally.

Banananaman91 committed:
Code review changes

Made changes based on code review feedback

Banananaman91 committed:
Extra change

Spotted one missed change

RubyNova reviewed (approved):
You're in the clear. Nice stuff!

Banananaman91 committed:
Merge branch 'Development' into feature/pathfinding
